src/foundry/common/abstract/type-data.d.mts

```diff
@@ -19,9 +22,53 @@ declare class _InternalTypeDataModel<
   BaseData extends Record<string, unknown> = Record<string, never>,
   DerivedData extends Record<string, unknown> = Record<string, never>,
   // This does not work if inlined. It's weird to put it here but it works.
-  _ComputedInstance extends object = Merge<BaseData, DerivedData>,
+  _ComputedInstance extends object = RemoveIndexSignatures<Merge<BaseData, DerivedData>>,
 > extends _InternalTypeDataModelConst<Schema, Parent, _ComputedInstance> {}
 
+// These properties are used to give a performant way of inferring the `BaseModel`, `BaseData` and `DerivedData` types.
+// This avoids checking the entire constraint `T extends TypeDataModel.Any` to infer out the `BaseData` and `DerivedData` types.
+// To prevent adding properties that could appear at runtime they are unique symbols.
+// This means that the only way to access them is to have a reference to the variables which will be impossible outside of this file.
+declare const __BaseModel: unique symbol;
+declare const __BaseData: unique symbol;
+declare const __DerivedData: unique symbol;
+
+declare namespace TypeDataModel {
+  export type Any = TypeDataModel<any, any, any, any>;
+
+  // This still is only allows classes descended from `TypeDataField` because these unique symbols aren't used elsewhere.
+  // These generic parameters seem to be required.
+  // This is likely because of a TypeScript bug in which concrete types like `any` or `unknown` aren't treated as carefully as a type parameter.
+  type TypeDataModelInternal<BaseModel, BaseData, DerivedData> = {
+    [__BaseModel]: BaseModel;
+    [__BaseData]: BaseData;
+    [__DerivedData]: DerivedData;
+  };
+
+  // Removes the base and derived data from the type.
+  // Has no extends bounds to simplify any checking logic.
+  type RemoveDerived<BaseThis, BaseModel, BaseData, DerivedData> = SimpleMerge<
+    Omit<BaseThis, keyof BaseData | keyof DerivedData>,
+    BaseModel
+  >;
+
+  export type PrepareBaseDataThis<BaseThis extends TypeDataModelInternal<any, any, any>> =
```

Comment:
I made these helper types so that users don't have to inline their complex definitions.